### PR TITLE
Don't charge duplicate fees for pending SCA payments

### LINKED ISSUE #6683: Unsuccessful SCA payments repeatedly introduce transaction fees (backoffice)
## Description

PR #6534 will introduce 3D Secure payments in the backoffice :tada:  

However, unsuccessful stripe SCA payment using 3D-secure cards "pile up" with their respective fees, on the order leading to incorrect values. 

For example, if the card numbers an Admin insert repeatedly fail due to:
- the authentication step (clicking the Fail Authentication) or
- invalid card for that transaction (ex., insufficient funds)

then we get the scenario shown under "Animated Gif/Screenshot".

This is a similar to #5818, but concerns backoffice orders.

## Expected Behavior

An uncusessful backoffice payment does not charge a transaction fee; only successful payments charge transaction fees. 

## Actual Behaviour

An uncusessful backoffice payment currently charges a transaction fee. 

## Steps to Reproduce

1. Add a Stripe-SCA payment in an order, in the backoffice -> Make sure there is a transaction fee associated with this payment method
2. Choose stripe-SCA as payment method and add a 3D secure card, like 4000 0027 6000 3184
3. When prompt click "Fail Authentication"
4. The payment should have failed, but notice the enterprise fee was charged, under "Order Details"
5. notice the bug is not occurring with non-3D secure cards like:

```
4000000000009995 | Charge is declined with a card_declined code. The decline_code attribute is insufficient_funds.
```

## Animated Gif/Screenshot

image

## Workaround

Refunding the incorrect amounts and manually removing the the transaction fees, in Adjustments. 

## Severity

bug-s3: a feature is broken but there is a workaround

## Your Environment

* Version used: not in production - related to #6534
* Browser name and version: Firefox 84
* Operating System and version (desktop or mobile): Desktop Ubuntu 20.04

### PR DESCRIPTION
**Pull request by andrewpbrett**
#### What? Why?

Close #6683
Here we mark an adjustment as ineligible if there SCA auth is required on a payment. Then when the customer completes SCA authorization, we process the payment intent and mark the adjustment as eligible again. 

#### What should we test?

Creating a payment that requires SCA authorization should not add a transaction fee to the order until the payment is authorized by the customer. 

#### Release notes

Fixed a bug where back office payments would create duplicate transaction fees each time a payment failed or needed authorization. 

Changelog Category: User facing changes

### CODE REVIEW

**Review by Matt-Yorkley — APPROVED** [review]

**Comment by Matt-Yorkley** [comment]
LGTM, but it probably needs some specs.

File: spec/services/process_payment_intent_spec.rb [hunk 1]
```diff
@@ -32,13 +32,15 @@
...
         allow(order).to receive(:deliver_order_confirmation_email)
+        payment.adjustment.update_attribute(:eligible, false)
       end
 
       it "completes the payment" do
         service.call!
         payment.reload
         expect(payment.state).to eq("completed")
         expect(payment.cvv_response_message).to be nil
+        expect(payment.adjustment.eligible).to be true
```
Comment (andrewpbrett) [inline spec/services/process_payment_intent_spec.rb:43]: I was originally doing 
`expect(payment).to have_received(:reinstate_adjustment_eligibility!)`
but it wasn't working as expected - it said it wasn't receiving the message, even though it was. 

So I went with this approach, which is a more useful test, although it does require reaching into the adjustment implementation just a little bit.

File: app/models/spree/payment.rb [hunk 2]
```diff
@@ -126,7 +126,9 @@ def payment_source
     end
 
     def ensure_correct_adjustment
-      revoke_adjustment_eligibility if ['failed', 'invalid'].include?(state)
+      if ['failed', 'invalid'].include?(state) || authorization_action_required?
```
Comment (sauloperez) [inline app/models/spree/payment.rb:129]: The extra complexity this conditional has and I think it'd be good the extract it, give it an abstract name, but I don't have any suggestion so I guess it's not a big deal. We'll see when a third condition comes into play.

**Review by sauloperez — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Hey @andrewpbrett ,

This is a tricky one... It looks like a step in the right direction, but I'm not sure it's ready for merging.

It looks as if this PR introduces the enterprise fees on the orders, after the payments take place, in both back- and front office. In the videos below we have the backoffice in the left side, and the customer's window on the right side.

After an enterprise fee of 3 eur is set to stripe-SCA payment method, and:

- **backoffice order** is placed (the waiting times are the times for the authentication email to arrive...)

Peek 2021-02-26 16-25

Notice, how the order is marked as "not paid" (non reglé), although the payment was successful.


- as a sanity test, **orders through the shopfront** were tested as well:

Peek 2021-02-26 15-58

In this case, the whole payment is charged (12 eur + enterprise fee of 3 eur) but the fee is not included in the order, so it appears as credit owed. 

This needs another go...

**Comment by andrewpbrett** [comment]
Thanks @filipefurtad0 - I can see why this PR wouldn't address enterprise fees that are a flat fee per order. I'm a little confused why it's actually removing them entirely however. I'll do some more digging and see if it warrants a separate issue.

**Comment by andrewpbrett** [comment]
@filipefurtad0 I took a look at this, and I think if I follow the steps you did, I get the same result, even on `master` and without doing anything related to SCA. The steps I followed were: 

1. Create a BO order for an order cycle with a per order enterprise fee
2. Don't click update/recalculate fees
3. Add customer details
4. Create a payment - note that the amount defaults to the amount without the fee
5. Complete the payment using a credit card that does not require SCA auth
6. Order is completed
7. Click Recalculate fees
8. Order is marked as Balance Due

Let me know if I misunderstood what you did or if that isn't a valid test in this case.

**Comment by filipefurtad0** [comment]
Hey @andrewpbrett ,

I realize this was perhaps not clear, from my testing notes before - thank you for reaching out! There is a spot on the previous notes, where "enterprise" is written but where "transaction" is meant. Sorry, this is corrected now.

The detailed steps you describe rather probe how/when enterprise fees included in the backoffice orders - these steps actually reproduce bug #5567.

So, to clarify: this specific PR would introduce (at the time of testing) a regression in which transaction fees would not be included in the order total, when checking out through the shopfront. This would be a considerable regression for the shopfront checkouts: it would render all Stripe orders placed though the shopfront in a "balance due" state. Maybe we need to cover this somehow with a feature spec.

Please note however, that this is known bug, for orders placed in the BO - see #5574.


So, the focus here is really on transaction fees, and on the steps to reproduce, on issue #6683 - currently in master. Using Stripe-SCA, in the backoffice:

- failed payment attempts with non-3D-auth. cards (like 4000000000000002) do not introduce transaction fees -> this is correct behavior. The payment did not succeed so a transaction fee was not changed

- failed payment attempts with 3D-authentication cards (like 4000008260003178, of by clicking "Fail Payment", in the auth. step) introduce a transaction fee, for each payment attempt -> this is incorrect, the transaction fee is charged regardless of the failed payment attempt 

I hope this helps.

**Comment by andrewpbrett** [comment]
Fixed in https://github.com/openfoodfoundation/openfoodnetwork/pull/6965

**Pull request closed without merging**